xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -672,14 +1048,268 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
       TF_RETURN_IF_ERROR(replace_with_new_cc(expectation, 1));
       TF_RETURN_IF_ERROR(replace_with_new_cc(instr, 2));
 
+      // Update the Custom Call associated with the input transpose orders.
+      transpose_orders.key() = new_custom_call->unique_id();
+      transpose_orders_.insert(std::move(transpose_orders));
+
       VLOG(1)
           << "Expectation and norm factor fused into layer norm Custom Call.";
     }
     return OkStatus();
   }
 
+  // Matches and rewrites the backward graph of layer norm patterns into Custom
+  // Calls to cuDNN when the associated forward graph has been rewritten into a
+  // cuDNN Custom Call.
+  Status MatchLayerNormBackward(HloInstruction* instr) {
```

Comment:
As in the forward pass, write the formula that is matched in the backwards pass, including the definitions of F1 and F2. Because F1 and F2 are used twice, you can define them above the main formula that's matched. E.g.

```
f = (variance + epsilon)^(-1/2)
Xcenter = X - E[X]
F1 = 2/n * (-1/2 * f**3) * sum(Xcenter * dy * scale) * Xcenter
...
```